### ISSUE TITLE
Better title needed:  Issue with Generics and Recursion?

### ISSUE THREAD

**Issue by SephReed** [microsoft/TypeScript#36878]
**TypeScript Version:**  3.7.5

**Search Terms:**
Generic recursion
Subtype
Also, already asked elsewhere

**Code**
```ts
export type TestProxy<TARGET> = (
  TARGET extends Array<any> ? TestArray<TARGET> : 
  TARGET
)

type TestArray<TYPE extends Array<any>> = {
  [key in (keyof TYPE & number)]: TYPE[key];
}
& Omit<TYPE, keyof Array<any>> // Special case for Array extensions
& {
  length: any, push: any, concat: any,
  join: any, reverse: any, shift: any, slice: any,
  sort: any, splice: any, unshift: any, indexOf: any,
  lastIndexOf: any, every: any, some: any, forEach: any,
  filter: any, reduce: any, reduceRight: any, find: any,
  findIndex: any, fill: any, copyWithin: any, [Symbol.iterator]: any,
  entries: any, keys: any, values: any, [Symbol.unscopables]: any,
  includes: any, flatMap: any, flat: any,

  // pop: any,  // CASE A everything works
  pop(): TestProxy<TYPE[number]>;  // CASE B 
  map<OUT>(cb: (it: TYPE[number], index: number, array: TestArray<TYPE>) => OUT): TestProxy<Array<OUT>>;
}

type TYPE = Array<string> & { extraField: string };
const thing: TYPE = null as any as TestProxy<TYPE>;  // Works in CASE A
const hat: string = "hat" as TestProxy<"hat">; // Always works

function proxyMe<TYPE>(target: TYPE): TestProxy<TYPE> {
  return target as any; // imagine proxy
}

const obj: { arr: number[][] } = { arr: [] };
const testArr = proxyMe([[1], [2], [3]]);
obj.arr = testArr;  // CASE A: works -- CASE B: Type 'TestArray<U[]>' is not assignable to type 'U[]
```

**Expected behavior:**
In Case A, it is valid to say `TYPE === TestProxy<TYPE>` for all types.  Therefore, having `pop()` return `TestProxy<TYPE>` in Case B should be valid.

**Actual behavior:**
In Case A, everything works.

In Case B, this error arises (see playground link)
> Type 'TestArray<number[][]>' is not assignable to type 'number[][]'.
  The types returned by 'pop()' are incompatible between these types.
    Type 'TestArray<number[]> | undefined' is not assignable to type 'number[] | undefined'.
      Type 'TestArray<number[]>' is not assignable to type 'number[] | undefined'.
        Type 'TestArray<number[]>' is not assignable to type 'number[]'.
          The types returned by 'map(...)' are incompatible between these types.
            Type 'TestArray<U[]>' is not assignable to type 'U[]'

**Playground Link:** 
[Playground Link](https://www.typescriptlang.org/play/#code/KYDwDg9gTgLgBDAnmYcAqwDOMAKUIiIA8aAggEoDiAomgHxwC8cAFAFBzoU1pygzAAdgBNMcUlCgBDYlMGIGAfnRYYE6cTJVaDAFxwOXbWjYBKNmyQoV2dTJIBNHNT4gBIsXdny6DZgG9DAG0Aa2BEOABLQVYwxAgAM3QnFwAyOEEAVwBbACNgKFMAXX00FNDwooBuNgBfNnSAeWzImEdnABo4OMTxSXs5BQYAemG4AGUUAGNIqQAbOCmpTFQE6D6NV3dMSIhBTAa4QM45oQBzGAALfUGusEzMa7hbxb2lmBv5DsMAKwhoz6ILpQYAANwKK0BXUekQSH2eXzgmDmkSmwChhkw0HhL0wYBRaKhcEy+0usJxiOiwlAjQSGJOyxgAEkRDS6QigXxwVBEESsdl0RyumsoNQpFMnrdDAlInMBFAiSDhJlCUK4EqVcByJEzpcKZyZSJ6XBDcIWdSQESZXM5kSphAwIgAOqtMmCIlBcaIPIQOYAOlaBSkMGgJSFhiEMCgkSwRLimCJoPmmVjas93tyvr9JMw9rAUlyp0wYalnGiUzmmWpCbVCTmwYAslIwFb6-rvoZRnBIC21XAuwBhUjjFykLkFRBXaJnOAAd2gIQOnB7LFMpVUeAImnKWTyBSKdCqnEHw5cACEDJxss2iI0AKr0FhTXL6FitUo7nL5KBFLpU0D6Lu35dFI-S8jYahge01B0KYTAMPeaBrhBm6EEQXi3g+vg1PUljIKgZTOEwGz2Ng0aCGcDDpP4WzSAAYjGczCPoZHTnAtQ1Pa+zwFOFEfkRzBZDazxiIMIkofgaGETBR79mMToLmI0RwEOI7iGwXHYHAlzBixUZscwABEOkwIZ4kYNgqHEMZwaGYecniHMs4yGI85QIuFgJCSUwwLsMRgJJiANsA0F0CwMCgWcwDwtJyEWbggWhUchggjAmRQDEEVQFF8DLByR5dpE15nNEqABVudQWJp8AQLkPz6DRoEKhkX4FEERTtexxGNZI+idRxGl7FpAi2JIxHlYQwUsEEQQAIy-nAQQAEwLUEADMRRFKYNS1T8fpNcRI2QVAsknmppD6G5i5wAAtDdKmnnAZ6lPhcAAOTxRhd7tXQb1RGIggQLlmA7GcggFqcCAQAgr1vd9RRAA)

**Related Issues:**
I've heard there are limitations to Generics that have been found elsewhere

**Comment by RyanCavanaugh** [comment]
Please ping me if you can supply a simpler repro; we generally cannot handle intricate and log examples like this in a timely manner

**Comment by SephReed** [comment]
I've got a simpler repro here:

```
type ARR<TYPE> = {
  pop(): TYPE;
  map<OUT>(cb: () => OUT): ARR<OUT>
  [key: number]: TYPE;
};

type Proxied<TARGET> = (
  TARGET extends ARR<any> ? ProxiedArray<TARGET> : 
  TARGET
)

type ProxiedArray<TYPE extends ARR<any>> =
  Omit<TYPE, "pop" | "map"> // don't want overloads
& {
  // pop(): any; // CASE A : works
  pop(): Proxied<TYPE[number]> // CASE B : fails
  map<OUT>(cb: () => OUT): Proxied<ARR<OUT>>;
}

const hat: string = "hat" as Proxied<string>; // Always works

type TYPE = ARR<string> & { extraField: string };
const thing: TYPE = null as any as Proxied<TYPE>;  // Works in CASE A


const testArr = {} as Proxied<ARR<ARR<string>>>;
testArr.pop(); // ProxiedArray

const obj = {} as { prox: ARR<ARR<string>> } ;
obj.prox = testArr;  // CASE A: works -- CASE B: Type 'ProxiedArray<ARR<OUT>>' is not assignable to type 'ARR<OUT>'
```

[Playground Link](https://www.typescriptlang.org/play/?ssl=30&ssc=107&pln=30&pc=118#code/C4TwDgpgBAggSnAPAFQJoAUCiA+KBeKAbwCgoowB7MACgEoAuKNLAblKgFsBDMRAeQCqybNQDGAI0Z18uQcgawE-IdnYBtANYQQjAHYBXDuIgAnALqNmmNgF82xUJCjoTFAB4BLCABMU8AOKYwvhQ1OzIAUFQEG7AELreAM6KSFy6ILgA-M6unj4wJiZcIH5wgcGM4ZHIxLTEDuDQLu5e3gVFJVbRsfFJKYhpGbh47HwcHsAoGJgANFAARJRg81AAPgvcy7gA9NtQAFL6icBQ3hS6AOQnADYeWlCJEBAeugDmUBQAbqbXFFxJxAAZER2LtyFQ6IxBiwoGCAMIwADKmFgUEYAHcKCYNIl2EtITkWj4plg1AYjKYzDs9gjkVAAEJoqAAMy4HmuuLIm2UwjEklCtBkUDkCmaeV88CQcmw2Fs9VE52OUAAFlxgIxjiYXu8CPNVcAVlxkmLWohNdrZbC9jBrujislMdjcQ0nF0CJKzcAtW9cMDCN0vVwAGJea7eDVe7VQOzEBW6JXAZXayzTEIGa7XKBGrPpLPG3KmqyWq1QADqWJxUBeUFpKJg9VjipOcWO7RChBsecJ4sQHo95p9MrYLeA7QAdPjaDCwSb8oVivKmx9xAArdud7P+sC5Rh9pQD14y6NQNgUVcT3IhEftGEl2uwDEV5IAWmfNaRKPplkaUAus7a84lB60rYBcVbJLoFAnEaiQeK8uhcOI1zQMAFBQI40AXMBKgXEAA)

**Comment by tadhgmister** [comment]
A possibly clearer [demo with playground](https://www.typescriptlang.org/play/index.html?ssl=1&ssc=1&pln=42&pc=3#code/LAKALgngDgpgBAFRgZzABQE4HsAeEA8CAggEoDiAoggHxwC8iplCcMOYMAdgCbJxEYMAQwJDOEWgH44AWSwBjANYDhBYuSq0AXIw0IA3KFAB6AFSm4Ad2FRYGOADMs9oYJFWAlmAAWcIY5hLOHkAV1QsAFs4LAA3GEEPbhQAGjgPTg4MtL4wLDgAI3gHEIAbEohgyKghMA98kvhLL19OZwihEr83CtNjUEhYWQVlbvwAeQwPAHNWdi5eflGxCVoGUDgNuAAyODG4hKTkccmpgG1OEIjCjABdWmNjaP3Jw7gobBjEmG51zZ2xiJeY7TVKKGAQLAOXbPL5HZbUe6PDqdZzTdIdRweGAlBZseQwKBgJ7xF4oX4bHYAb1OAGUIFcsCUAHReeI1Zw3HTLfRwWn0-KMpkhTjIeRYar1FCcvziHkAXzgDzgyHS+N2gKJHWQeSmMDAfBe0ShyH5jLgYIgfEs8E4MG+cB8HhyeQwMHa6UqEWqtUlTKMIHSHAwDiEar2JNhhFolPJb3FAAoAJQ6JCoTC4NS0AA+cGFSQc6W+hhAmzg7Sg4wAqjR4-J8jp49i3SnUukkjgdBcrvFUq4MDpHchE-RaGNq8mhkoVCIqwhTndi3LQEl5CVXEVhfJalhOG9sHgZDBCABNNAUajxsCuXVgFOnigT1Pofdqe-Uf3FThbjw75XeLCWBQggAOrNCoZBcPEHjyPgRCzJkCzLPOF4cKgKg6HIU6jEQ1CJjGJabEqO7lFY3hcHGUDZJ6ERcBw3DRCERI+PAJLOH+AF8GwYDCHAJSFn4AqMXAADkmEjKosHUMJVFCMgKpTJwQiSg6eTMZUIrcUIgZGg60DwMJRDCbGYoaXArQYO0JTofw9AOigYAqIuiqmKACB6SJYnTgQOHSU6ZlYJqcnTIpym5LpgwGcJfogL0H6btuu7IP+gEgc0Yz5AAVlGl72elGUpvZ6Z4FGeHGTuqD+RZHR5SmtmoWAeVOWYrnucJT5FZmvl8K0gXySFDQqeF+kIFFvxtYVL5Rl1fhBQpSkDWFakmagwjaZCQ0iSNqT5EJbXSWKpT0YUaQaWItQ1PaTQ+H4cDcB4DgOPEtHKiE+QDPA63LZp2nCZScqjQRGxuRFLA5p5ow0NNPUzX183wItrUjdFpYbG1cBg8MXlTTJs39fDqlkepK1aVk63vZtwnbbtI2VIdBTwOkqBnR4F30Vdvj+HdD1PVkyCveTn3ld9WS-f9yMo0DrXgxJkNUdDsmw6FqmIwDEulqJmMQ1JOOKwtBPwF9q2k1C5N7VTRJ7bTOL0ydTMZCzdGeNdnP3Y9rq8-z7mCxpRsW39UX6vGABMADMQdB4moCxSAn7fr+SUAUBGCgT4KjZfV1nSzOc53KVgOKo8g5vArKBpFC4YHCgTJQAmw5+WKEQ0Rk9pYIxZWmeZlnWTndX2Y5oAKs1IDA-pWdqMhUMBTDwVw4Nps56riCE+9Bp6iEGC2kdFTCTXUBJtJ64nQ33p1ANhRgNa5HMcg8N6cg4uSyDE0ZlG6O5jwMAFpvk+9TPSsbWjHMeZP6Fm4IvUsI9NrP2KrLPy8tcazwRiDN+wCv7fHASjSBgDJziWztrOBU8FZ-z1gA0G798ygIwRLLBY9sYEN-nNf+88UEfzQWAh+atH6j01jLfB3VCEIKYSrDhnDUa0NgXwIhjCSFLSFr7HS89KYFGpvtVu1tjqMyvPbVmTsOa3VdjzIkfM3peyhIbEmfsxaxlEZsLB7VJoSP8gwvGc9hHWJsZscaaYHF8OntI-GDpCbmLWibFW5sKZWyOgzU62jHbsxulzN2z1jECzMXIixIl-YiI8RAxGb9xG+PgbrAJijsk5PVmQgp00pEuMWkE9JITSFKJ2hbGmB11HRLtudOJzQEkGPdkYz2gxvbEx+lk9x5TMFSx4Xgn+fjanKxBlQyZasNZYV4dUwRMj6k+wyWTMJyjWmqLphomJ3TLq9JdtzAZL0THDLSbssZYtA6h3DpHGKxggA)

```typescript
type TestProxy<TARGET> = TARGET extends Array<any> ? MockArray<TARGET> : TARGET;

/** wrapper for array with a few custom overrides, intent is to be fully compatible with normal array */
type MockArray<Orig extends Array<any>> = 
    & Overrides<Orig[number]> // overrides provided
    & Omit<Orig, keyof Overrides<any>> // all original fields except overrides
    & {[Symbol.iterator]: any; [Symbol.unscopables]: any; } // since Omit also gets rid of symbol keys we need this to remain compatible.

interface Overrides<T> {
    pop(): TestProxy<T> | undefined;
    map<OUT>(cb: (elem: T, index: number, arr: this) => OUT): MockArray<OUT[]>;
}
function showErrWithArrGeneric<A extends any[]>(testArr: MockArray<A>){
    // only when pop is commented out the error shows extra line about 'MockArray<A>' is assignable to the constraint of type 'A'
    const normalArr: A = testArr;
} /*
Type 'MockArray<A>' is not assignable to type 'A'.
*/

function showErrWithObj<T>(testObj: TestProxy<T>){
    const normalObj: T = testObj;
} /*
Type 'TestProxy<T>' is not assignable to type 'T'.
  'TestProxy<T>' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint '{}'.
    Type 'T | MockArray<T>' is not assignable to type 'T'.
      'T | MockArray<T>' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint '{}'.
        Type 'MockArray<T>' is not assignable to type 'T'.
          'MockArray<T>' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint '{}'.ts(2322)
*/
function showErrWithArr<T>(testArr: MockArray<T[]>){
    // this passes if Overrides.pop() is commented out
    const normalArr: T[] = testArr;
} /*
Type 'MockArray<T[]>' is not assignable to type 'T[]'.
  The types returned by 'pop()' are incompatible between these types.
    Type 'TestProxy<T> | undefined' is not assignable to type 'T | undefined'.
      Type 'TestProxy<T>' is not assignable to type 'T | undefined'.
        Type 'T | MockArray<T>' is not assignable to type 'T | undefined'.
          Type 'MockArray<T>' is not assignable to type 'T | undefined'.
            Type 'MockArray<T>' is not assignable to type 'T'.
              'MockArray<T>' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint '{}'.
                Type 'TestProxy<T>' is not assignable to type 'T'.
                  'TestProxy<T>' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint '{}'.
                    Type 'T | MockArray<T>' is not assignable to type 'T'.
                      'T | MockArray<T>' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint '{}'.
                        Type 'MockArray<T>' is not assignable to type 'T'.
                          'MockArray<T>' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint '{}'.ts(2322)
*/
```
When the `map` override is commented out concrete arrays work fine since there is no wildcard generics anywhere and typescript can figure out that all is fine:

```typescript
interface Overrides<T> {
    pop(): TestProxy<T> | undefined;
    // map<OUT>(cb: (elem: T, index: number, arr: this) => OUT): MockArray<OUT[]>;
}
declare function proxyMe<TYPE>(target: TYPE): TestProxy<TYPE>
const a: number[] = proxyMe([1,2,3]) // works fine
```
When `pop` override is commented out then `MockArray<T[]>` is assignable to `T[]` since typescript knows there aren't any weird extensions on an array so the third case passes.

```typescript
interface Overrides<T> {
    //pop(): TestProxy<T> | undefined;
    map<OUT>(cb: (elem: T, index: number, arr: this) => OUT): MockArray<OUT[]>;
}
function WORKS_NOW<T>(testArr: MockArray<T[]>){
    const normalArr: T[] = testArr;
} 
```

So I know the root error is that `MockArray<A>` is not assignable to `A` for wildcard generics because `A` as a total wildcard generic could defines extra overloads for `pop` or `map` then we'd omit them in our mock so it's not unconditionally assignable.  It is still strange we get the error message with repeating lines though.

Hope this helps :)

**Comment by SephReed** [comment]
That makes sense, and it feels good to finally understand why it's not 100% valid typing.

If I understand correctly, an exact type (https://github.com/microsoft/TypeScript/issues/12936) would be able to solve this by defining the variable as non-extendable, correct?

**Comment by SephReed** [comment]
Can you think of a way to write:

```
type TestProxy<TARGET> = TARGET extends Array<any> ? MockArray<TARGET> : TARGET;
```

such that TARGET will only become a MockArray if it is not an extension?

**Comment by tadhgmister** [comment]
> Can you think of a way to write:
> 
> ```ts
> type TestProxy<TARGET> = TARGET extends Array<any> ? MockArray<TARGET> : TARGET;
> ```
> 
> such that TARGET will only become a MockArray if it is exactly an array and not an extension?

you can by somehow using `Array<any> extends TARGET` but I'm pretty sure when typescript puts a deferred generic through one of these conditional types it can only statically know that the result is a union of all possible branches, so no matter how you setup the conditional type a wildcard generic will still be seen as `T | TestArray<T>` and therefore only pass if `TestArray<T>` is unconditionally assignable to `T`, so I'm afraid to satisfy typescript you need to use an intersection with the underlying type.  I think my [suggestion here](https://github.com/microsoft/TypeScript/issues/36821#issuecomment-588511452) is as good as you'll get, the only improvement might be to have the overrides extend `Array<E>` so it will catch you if your overrides are blatantly incompatible with a standard array
```typescript
type TestProxy<T> = T extends Array<any> ? TestArray<T> : T;
type TestArray<Arr extends any[]> = ArrOverrides<Arr[number]> & Arr;
interface ArrOverrides<E> extends Array<E>{
/* Interface 'ArrOverrides<E>' incorrectly extends interface 'E[]'.
     The types returned by 'find(...)' are incompatible between these types.
       Type 'string' is not assignable to type 'E | undefined'.ts(2430) */
    find(): string;
}
```

If you really wanted to put some extra checks to not allow weird extra fields then you could use `any[] extends Arr` inside the TestArray definition like this:

```typescript
type TestArrayValidating<Arr extends any[]> = any[] extends Arr ? ArrOverrides<Arr[number]> & Arr : never;
type Valid = TestArrayValidating<number[]>
type Invalid = TestArrayValidating<number[] & {extraField: string}>;
type MaybeShouldBeValidButIsnt = TestArrayValidating<Valid> // passing a TestArray ends up never
```
When typescript puts a generic through this, it doesn't know if `any[]` is assignable to it so it takes the union of both branches, but since one of the branches is `never` the union is the same as just the first branch so this is just as compatible with wildcard (deferred) generics as the first definition, and if you expect people to add weird extra fields and want to end up breaking their types then this might be useful? I feel like wanting `TestArray<TestArray<number[]>>` to still be valid would be better than wanting extra fields to cause it all to end up `never`.

Anyway, good luck with your [really dope shit](https://github.com/microsoft/TypeScript/issues/36821#issuecomment-588408856) 😃

**Comment by SephReed** [comment]
Thank you for all your patience.

One last thing:  if an `Exact<TYPE>` ever comes into existence, it should -- theoretically -- make it so the equivalence I'm trying to achieve is possible, right?

**Comment by tadhgmister** [comment]
> One last thing: if an `Exact<TYPE>` ever comes into existence, it should -- theoretically -- make it so the equivalence I'm trying to achieve is possible, right?

Yes probably, I have something close implemented below. But I'm going to make an argument for why what you are trying to achieve isn't actually what you want. (And why this is true for pretty much everyone looking for an `Exact<TYPE>`which is why it is never going to get added to typescript)

I brought up `any[] & {extraField: any}` as a counter example when you were saying things like "provably an error in the compiler", and once you realize that that is the kind of case that typescript is complaining about it is natural to want to specify a type contract that makes that illegal to use in your code.

What you don't realize that you are more likely to run into reasonable cases of "array extensions" than unreasonable ones. Like if you pass an array with an extra field to a function called `proxyMe` which is designed for normal arrays, you shouldn't get your hopes up that it preserves your unusual behaviour.  However the following 2 lines of code is very reasonable to expect to work for any (valid array) type of `MyArr`:

```typescript
declare const testArr: TestArray<MyArr>;
const x: MyArr[0] = testArr[0]; // first element of MyArr assigned to first element of wrapper
```
Normal arrays have an index `number` not a specific index for `0`, **but array literals do:**

```ts
type MyArr = ["a", "b", "c"];
type ExplicitEquivelent = Array<"a"|"b"|"c"> & {0: "a", 1:"b", 2:"c", length: 3}
```

This is a valid extension of the basic array type that you almost certainly already support, and the definition of `TestArray` that always uses an intersection with the original type will always preserve this. If we just want to ensure that the type is a plain array then we end up losing information and this ends up failing:

```typescript
type EnsureRegularArray<T> = T extends Array<infer E> ? Array<E> : T;

declare const testArr2: EnsureRegularArray<MyArr>;
const y: MyArr[0] = testArr2[0];  // Error: Type '"a" | "b" | "c"' is not assignable to type '"a"'.
```

If your case primarily revolves around the `pop` method I doubt array literals are an expected use case, but I doubt you want to deny them just because they are "more specific".

Now lets talk about how typing relates to runtime,  lets say we did have syntax of this form:
```typescript
type DreamProxy<T> = T isMutuallyAssignableWith Array<infer E> ? OverrideArr<E> : T
```
 In this case the type checker would know that the first branch is only visited if `Array<E>` is assignable to `T` so we just need to validate that `OverrideArr<E>` is assignable to `Array<E>` (which it is because we extended Array) so this would play nice with wildcard generics. Yay!

Here's the catch, if we implement `proxyMe` to check `obj instanceof Array` to actually check if we are using an array, then it means that **an array extension will be proxied at runtime** even though **the type will claim to return the same object.**  I'm pretty sure this is the primary argument against the `Exact<TYPE>`, because **most of the time you can't actually enforce at runtime that only an exact type signature will be converted.**  If you still have any doubts try running [this playground](https://www.typescriptlang.org/play/index.html#code/PTAEGMBsHsGcFNYBdRIBYEsB2BzV1QBbeLJDaLeAE1CvgDNsMyKAaUAKBFAsgE9QODADdEoAIIAnSQHlR0jHVigAPAFEAfKAz0J0gIYD1WjMsIBXJOf2R+ofbFgYcWfQCNI8fKAAq9rDSS8FaSWMp+2P4C0OjwkhAO8AB0XGBYMV7o+iixoAAGAGQ+eaCwaNDmkFRYAOQo+vT08OD1WHzo2DjszDXKAFbmyPagAO6GPLo4wWS49pIVAamofAAOiOCSGCtIypAYhMzZ5FioaNnaOZhYANbKWZfnbpL6WOBooxVVtShuXg5OLncnm8Pg4SFWXgAIkF9IQAArzAAefBUPi0AF5fKB4IikCQqMopM8UdgmvFNKAAPygAAURMMxmxuPx4Spelk8k2SkZBSxAC5fABKUACnwAbiW6Txp3OuRW0GwKGgug6yjohAIpm8ZWgIxlly10PgsIR0GRyzW2mUkXo+haElAADkAJIAcQAEj4ALLiABKaiWv3A+kGXhGXjOolA6VG4yQBBI9GgknAfxQknMpH2mTOSqwdgzJ2Vpy8B1sc2JyhGnxov1AKyRGGoHA49EzLWO9aRfC98FRGhp0DcfVFgoFRpN3f7oAA3hxQAvtLpB8PtGEkC9U8X6XxhXPFwfQNxxI6AJoSX2+8TnmQAIQAUmoAMJ+ADqzoAMh-QLe1KA4b6MgABrOmokLsAAyjIoCvu655+n+r5-k+H7iM6XrOo6rqgDIPjumovrvhBf6YaAE7wt2lLzoeC5BCEJyUHqRJyHEXKIDSSScUOfTCg4ZEwhRZoomiEoHgAvtikAILO1E0XR5ihDwq58eRprIv2omLmJHDaVA-zsixCjchSOJ4gEhIGCimhzrJ9bQCsNJ7rZB7yYpDZCb2NKwOYaySEk8oOYKgqaQu2kHoQ+grCoMgAKpojS4BuAKNLMAKajdAEOIClg5iEL8kjsPollpQA2gAusK6JaLFPhOTRtHBApJzucinneb5SQRQ5iVBbZYWLvM4DXAA+hUSCOTJ9UQBQsDQJ4SQwDgNKquwABEzoQaAgFPgA0ph2E1QAhIdq3BX1OktuAM0oAAauIH7OpCw0nk9r4yL6O2bZiLU9vANIlQAjOwABM7AAMwVRKd0PU9L3DW9H0QUkg0jWNjlikeYB8IYh2XddF6+qAmKAyD4NlfYyhXeuEpU0MPi4fdH6nsN0OPcNMXEcNT7iJzABi73DQBwHnt93aeX6vpnfTPiM8zrNPRzahczzSv876guAUBp7I9AQ2jZY6OYx8ki3PY6aZmQxDsOCaywBsWwoKY3zTVgwjYKmVApHj66gB+agQRBLP3WzivK8RROgCVq36Kta1uLHoCreAq3k3xyCbLg5WgLyM5Mkgzy802VQCunnSgGJEp+wHQcw+znPc8RSSmQXRc0Jiq0QdAxC0Nk+igGgcTwKt3tDK+gFYRHP2eVXgfy3XSsN2owWcNwNvrJs2wQJA+j7HcmDWsowg2IooBuJYJZdkJtCKM71zwPAKwXBwtNzcki00qtzd94w8BVDwzXdnYCMTAbxRh-CCM7dq8pJAzDwKYAUCcx4yCwk3XELdf5UDOtwfQOAd5YHYCjfW41hTVhNsoc4hZLZeDPigcMNQozxj-mvO2G8lSxEkCMUw8AOBIJQYQtGgogA) since it should convince you that this notion of `Exact<TYPE>` is almost certainly not what you want.

```typescript
// closest thing to isMutuallyAssignableWith that we can get.
// only gives ArrOverrides <E> if Array <E> is mutually assignable to T and returns T in any other case.
// note that the `&T` shouldn't affect anything, it's just a way of getting around
// typescripts limitation that it thinks that that branch wouldn't be assignable to T
type DreamProxy<T> = T extends Array<infer E> ? (Array<E> extends T ? ArrOverrides<E> & T : T) : T;
// note that the point of this demo is to show that this DreamProxy type is in fact A NIGHTMARE
// because we have no way to enforce at runtime that only run of the mill arrays would be proxied

function proxyMe<T>(obj: T): DreamProxy<T> {
    if (obj instanceof Array) {
        // ANY ARRAY OBJECT WILL BE PROXIED, SO WHY ARE WE CLAIMING OTHERWISE IN DreamProxy?
        return new ArrOverrides(...obj) as DreamProxy<T>;
    } else {
        return obj as DreamProxy<T>;
    }
}
class ArrOverrides<E> extends Array<E>{

    pop() {
        return proxyMe(super.pop());
    }
    map<OUT>(cb: (it: E, index: number, array: E[]) => OUT) {
        return proxyMe(super.map(cb))
    }
    rock_out() {
        console.log(this, "IS ROCKING OUT!!");
    }
}

const VALID_AND_WORKS = proxyMe([1, 2, 3]);
VALID_AND_WORKS.rock_out(); // yay!

const ARR = [1, 2, 3] as const;
const TOTALLY_VALID_USE_CASE_FOR_PROXY = proxyMe(ARR);
TOTALLY_VALID_USE_CASE_FOR_PROXY.rock_out(); // works at runtime, typescript isn't convinced.


const LESS_VALID_USE_CASE = ["a", "b", "c"] as string[] & { extraField: string };
LESS_VALID_USE_CASE.extraField = "Some data here"

const WRONG = proxyMe(LESS_VALID_USE_CASE); 
// typescript claims this is valid but the proxy didn't keep it
console.log("extra field on proxy, which we aren't supporting is: ", WRONG.extraField);
// again, rock_out() works at runtime but we've told typescript otherwise
WRONG.rock_out()
```

So yeah, `DreamProxy` is valid, the `&T` part in it is to get around a limitation in typescript and should not affect the effective types at all.  **There is absolutely no way to enforce it at runtime.**

**Comment by SephReed** [comment]
I see why DreamProxy doesn't exist.  And, for my use case, it's totally okay if extended and const arrays are blocked, but the typing on DreamProxy seems nightmarishly confusing for any user who'd get an error there.

-------

At this point, I feel content to have seen what seems to be every possible option.  Your first suggestion is the best.

----

It still seems like there should be a way to do this.  This type exists, I've created it.  I just don't know how to explain it.

**Comment by RyanCavanaugh** [comment]
@SephReed is there a TL;DR version of something actionable we can look at, or can I mark this as `Question` ?

**Comment by SephReed** [comment]
tl;dr  It is impossible to describe this type such that it is equivalent to the object it is proxying:

```
type Impossible<TYPE> = ???;

let numberOfTimesPopHasBeenUsed = 0;
function proxyAnything<TYPE>(target: TYPE): Impossible<TYPE> {
  if (Array.isArray(target) === false) { return target;  }

  return new Proxy(target, {
    get: (_, propName) => {
      if (propName === "pop") { 
        numberOfTimesPopHasBeenUsed++;
        return () => proxyAnything(target.pop()); 
      }
      return target[propName];
    }
  });
}
```

**Comment by tadhgmister** [comment]
wait... @SephReed  if you are using a `Proxy` then you **DO** conform to the type `Overrides<..> & T`.  That is how proxies work, it conforms to the exact type of the target in addition to anything you override.  

In fact because you aren't changing any behaviour other than wrapping in proxies... you don't even need a separate type, **`type TestProxy<T> = T` accurately describes your proxy object.** Like the call signature of `Proxy` is `new<T>(target: T, handler:..) => T` - the type contract states that the proxy conforms exactly to the targets type, for all viable types.

if you just remove `: Impossible<TYPE>` from the call signature of `proxyAnything` typescript already infers that the return type is always `TYPE` because that is what `new Proxy` returns...

**Comment by SephReed** [comment]
Man, getting this boiled down is really, really hard.  I've added a `target` prop.  This proxy thing I'm working on is largely the same, but with a couple extra functions.  It's the same enough it should be assignable to the original type, but different enough it needs its own type.

**Comment by tadhgmister** [comment]
> is largely the same, but with a couple extra functions

Ok good, because if the solution was just "remove explicit return type from your proxy function" I would feel really silly taking so long to get there.

So then yes my first suggestion of `Overrides & T` precisely describes your case, if you did use a `any[] & {extraField: string}` then the extra field is preserved at runtime by the proxy and using an intersection with the original type tells typescript the same.

@RyanCavanaugh  [my first comment](#issuecomment-588532141) shows an error message that has the line `Type 'MockArray<T>' is not assignable to type 'T'` **repeated twice.**  If that seems weird then maybe it's actionable but I think everything else was question material.
